File: 2015/Day5/src/day5_ai.py
def has_three_vowels(s: str) -> bool:
    """Check if string has at least three vowels (aeiou)."""
    vowels = 'aeiou'
    return sum(1 for char in s if char in vowels) >= 3

def has_double_letter(s: str) -> bool:
    """Check if string has at least one letter appearing twice in a row."""
    for i in range(len(s) - 1):
        if s[i] == s[i + 1]:
            return True
    return False

def has_no_disallowed(s: str) -> bool:
    """Check if string does not contain ab, cd, pq, or xy."""
    disallowed = {'ab', 'cd', 'pq', 'xy'}
    for i in range(len(s) - 1):
        if s[i:i + 2] in disallowed:
            return False
    return True

def is_nice_part1(s: str) -> bool:
    """Check if string is nice under Part 1 rules."""
    return has_three_vowels(s) and has_double_letter(s) and has_no_disallowed(s)

def has_pair_twice(s: str) -> bool:
    """Check if string has a pair of letters appearing at least twice without overlap."""
    seen_pairs = {}
    for i in range(len(s) - 1):
        pair = s[i:i + 2]
        if pair in seen_pairs:
            # Check for overlap (pair must not start at the previous pair's end)
            if i > seen_pairs[pair] + 1:
                return True
        else:
            seen_pairs[pair] = i
    return False

def has_repeat_with_one_between(s: str) -> bool:
    """Check if string has a letter repeating with exactly one letter between."""
    for i in range(len(s) - 2):
        if s[i] == s[i + 2]:
            return True
    return False

def is_nice_part2(s: str) -> bool:
    """Check if string is nice under Part 2 rules."""
    return has_pair_twice(s) and has_repeat_with_one_between(s)

def count_nice_strings(filename: str) -> tuple[int, int]:
    """Count nice strings for both parts from the input file."""
    nice_count_part1 = 0
    nice_count_part2 = 0
    
    with open(filename, 'r') as file:
        for line in file:
            s = line.strip()
            if is_nice_part1(s):
                nice_count_part1 += 1
            if is_nice_part2(s):
                nice_count_part2 += 1
                
    return nice_count_part1, nice_count_part2

# Replace 'input.txt' with the path to your input file
filename = 'input.txt'
part1_count, part2_count = count_nice_strings(filename)
print(f"Part 1: {part1_count} nice strings")
print(f"Part 2: {part2_count} nice strings")

File: 2015/Day5/src/day5_ai.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <unordered_set>

bool is_nice_part1(const std::string& s) {
    int vowel_count = 0;
    bool has_double = false;
    std::unordered_set<std::string> forbidden = {"ab", "cd", "pq", "xy"};
    
    // Check for forbidden substrings
    for (size_t i = 0; i < s.length() - 1; ++i) {
        if (forbidden.count(s.substr(i, 2)) > 0) {
            return false;
        }
    }
    
    // Count vowels and check for double letters
    for (size_t i = 0; i < s.length(); ++i) {
        if (s[i] == 'a' || s[i] == 'e' || s[i] == 'i' || s[i] == 'o' || s[i] == 'u') {
            ++vowel_count;
        }
        if (i < s.length() - 1 && s[i] == s[i + 1]) {
            has_double = true;
        }
    }
    
    return vowel_count >= 3 && has_double;
}

bool is_nice_part2(const std::string& s) {
    bool has_pair = false;
    bool has_repeat = false;
    
    // Check for pair of two letters appearing twice without overlapping
    for (size_t i = 0; i < s.length() - 1 && !has_pair; ++i) {
        std::string pair = s.substr(i, 2);
        // Look for this pair in the rest of the string
        for (size_t j = i + 2; j < s.length() - 1; ++j) {
            if (s.substr(j, 2) == pair) {
                has_pair = true;
                break;
            }
        }
    }
    
    // Check for letter repeating with one letter between
    for (size_t i = 0; i < s.length() - 2; ++i) {
        if (s[i] == s[i + 2]) {
            has_repeat = true;
            break;
        }
    }
    
    return has_pair && has_repeat;
}

int main() {
    std::ifstream file("input.txt");
    if (!file) {
        std::cerr << "Cannot open input.txt" << std::endl;
        return 1;
    }
    
    std::vector<std::string> strings;
    std::string line;
    while (std::getline(file, line)) {
        if (!line.empty()) {
            strings.push_back(line);
        }
    }
    file.close();
    
    // Part 1
    int nice_count_part1 = 0;
    for (const auto& s : strings) {
        if (is_nice_part1(s)) {
            ++nice_count_part1;
        }
    }
    
    // Part 2
    int nice_count_part2 = 0;
    for (const auto& s : strings) {
        if (is_nice_part2(s)) {
            ++nice_count_part2;
        }
    }
    
    std::cout << "Part 1: " << nice_count_part1 << std::endl; // Expected: 255
    std::cout << "Part 2: " << nice_count_part2 << std::endl; // Expected: 55
    
    return 0;
}